File: js/subway.js
import { createScene } from './environment.js';
import { Agent } from './agent.js';
import { Wall } from './wall.js';
import { updateAgents } from './physics.js'
import * as UTILS from './utils.js'
import * as THREE from 'three';

let agents = [];
let walls = [];
let timestep;

const CONFIG = {
    COUNT: 20,
    RADIUS: 1,
    MAXSPEED: 5,
    MAXFORCE: 30,
    HORIZON: 5,
    K: 2,
    AVOID: 1,
    SIDESTEP: 5,
}

const gaps = [-33, -11, 11, 33];  
let wallTargets = [];
let agentsTargets = [];
let assigned = [];
let floor;
let animationDone = false;

const group1Mat = new THREE.MeshLambertMaterial({
    color: 0x00ff00
});
const group2Mat = new THREE.MeshLambertMaterial({
    color: 0xff0000
});

const { renderer, scene, camera } = createScene();
init();

function nearestDoor(z) {
    return gaps.sort((a, b) => Math.abs(z - a) - Math.abs(z - b))[0];
}

function init() {
    const wallData = [
        [100, 4, false, new THREE.Vector3(50, 2, 0)],
        [15, 4, true, new THREE.Vector3(42.5, 2, -50)],
        [15, 4, true, new THREE.Vector3(42.5, 2, 50)],
        [12, 4, false, new THREE.Vector3(35, 2, -44)],
        [12, 4, false, new THREE.Vector3(35, 2, -22)],
        [12, 4, false, new THREE.Vector3(35, 2, 0)],
        [12, 4, false, new THREE.Vector3(35, 2, 22)],
        [12, 4, false, new THREE.Vector3(35, 2, 44)],
    ];

    wallData.forEach(([width, height, vertical, position]) => {
        position.z += 200;
        const wall = new Wall(width, height, vertical, position);
        scene.add(wall.mesh);
        walls.push(wall);
        wallTargets.push(new THREE.Vector3(wall.mesh.position.x, 2, wall.mesh.position.z - 200));
    });

    const floorMaterial = new THREE.MeshPhongMaterial({ color: 0x222222, side: THREE.DoubleSide });

    const geometry9 = new THREE.PlaneGeometry(100, 15);
    const plane9 = new THREE.Mesh(geometry9, floorMaterial);
    plane9.castShadow = true;
    plane9.receiveShadow = true;    
    plane9.rotation.set(Math.PI / 2, 0, Math.PI / 2);
    plane9.position.set(42.5, 0.05, 200);
    floor = plane9;
    scene.add(plane9);

    for (let i = 0; i < CONFIG.COUNT; i++) {
        const pos = UTILS.getPosition(40, 47.5, -39, 39);
        const maxSpeed = Math.random() * (CONFIG.MAXSPEED - 5) + 5;

        agents.push(new Agent(
            i,
            pos[0], 2, pos[1],
            0, 0, 0,
            0, 0, 0,
            0, 2, 0,
            CONFIG.RADIUS, maxSpeed, CONFIG.MAXFORCE, CONFIG.HORIZON, 
            CONFIG.K, CONFIG.AVOID, CONFIG.SIDESTEP
        ));
        agents[i].setData("group", 1);
        agentsTargets.push(new THREE.Vector3(agents[i].position.x, 2, agents[i].position.z));
        agents[i].position.z += 200;
        assigned.push(false);
    }

    for (let i = 0; i < CONFIG.COUNT; i++) {
        const pos = UTILS.getPosition(0, 23, -39, 39);
        const maxSpeed = Math.random() * (CONFIG.MAXSPEED - 5) + 5;

        agents.push(new Agent(
            i + CONFIG.COUNT,
            pos[0], 2, pos[1],
            0, 0, 0,
            0, 0, 0,
            0, 2, 0,
            CONFIG.RADIUS, maxSpeed, CONFIG.MAXFORCE, CONFIG.HORIZON,
            CONFIG.K, CONFIG.AVOID, CONFIG.SIDESTEP,
        ));
        agents[i + CONFIG.COUNT].setData("group", 2);
    }

    for (let i = 0; i < CONFIG.COUNT; i++) {
        const agentGeometry1 = new THREE.CylinderGeometry(agents[i].radius, 1, 4, 16);
        const agent1 = new THREE.Mesh(agentGeometry1, group1Mat);
        agent1.castShadow = true;
        agent1.receiveShadow = true;
        scene.add(agent1);
        agents[i].setData("agent", agent1);

        const agentGeometry2 = new THREE.CylinderGeometry(agents[i + CONFIG.COUNT].radius, 1, 4, 16);
        const agent2 = new THREE.Mesh(agentGeometry2, group2Mat);
        agent2.castShadow = true;
        agent2.receiveShadow = true;
        scene.add(agent2);
        agents[i + CONFIG.COUNT].setData("agent", agent2);
    }
}

function animate() {
    requestAnimationFrame(animate);
    
    // subway arriving at station animation
    if (!animationDone) {
        let done = true;

        // smoothly lerp walls 
        for (let i = 0; i < walls.length; i++) {
            walls[i].mesh.position.lerp(wallTargets[i], 0.005);

            if (walls[i].mesh.position.distanceTo(wallTargets[i]) > 0.1) {
                done = false;
            } else {
                walls[i].box.setFromObject(walls[i].mesh);
            }
        }

        // smoothly lerp floor
        floor.position.lerp(new THREE.Vector3(42.5, 0.05, 0), 0.005);
        if (floor.position.distanceTo(new THREE.Vector3(42.5, 0.05, 0)) > 0.1) {
            done = false;
        }

        // smoothly lerp agents
        for (let i = 0; i < agents.length; i++) {
            if (agents[i].getData("group") == 1) {
                agents[i].position.lerp(agentsTargets[i], 0.005);

                if (agents[i].position.distanceTo(agentsTargets[i]) > 0.1) {
                    done = false;
                }
            }
            agents[i].getData("agent").position.copy(agents[i].position);
            agents[i].getData("agent").material = agents[i].getData("group") == 1 ? group1Mat : group2Mat;
        }

        if (done) {
            animationDone = true;
        }

        renderer.render(scene, camera);
        return;
    }

    agents.forEach(function(member) {
        // navigate agents to subway doors and outside
        switch (member.getData("group")) {
            case 1:
                if (member.position.x > 35 + CONFIG.RADIUS && !assigned[member.id]) {
                    member.target.z = nearestDoor(member.position.z);
                    member.target.x = 35 - 0.5;
                } else if (!assigned[member.id]) {
                    [member.target.x, member.target.z] = UTILS.getPosition(0, 23, -39, 39);
                    assigned[member.id] = true;
                }
                break;
                
            case 2:
                if (member.position.x < 35 - CONFIG.RADIUS && !assigned[member.id]) {
                    member.target.z = nearestDoor(member.position.z);
                    member.target.x = 35 + 0.5;
                } else if (!assigned[member.id]) {
                    [member.target.x, member.target.z] = UTILS.getPosition(40, 47.5, -39, 39);
                    assigned[member.id] = true;
                } 
                break;
        }
    });

    timestep = document.getElementById("timestep").value;
    document.getElementById("timestepValue").innerHTML = timestep;
    agents.forEach(function(member) {
        updateAgents(member, agents, timestep);
    });

    agents.forEach(function(agent) {
        walls.forEach(function(wall) {
            wall.collisionResolve(agent);
        });
    });

    agents.forEach(function(member) {
        member.getData("agent").position.copy(member.position);
        member.getData("agent").material = member.getData("group") == 1 ? group1Mat : group2Mat;
    });

    renderer.render(scene, camera);
};

animate();

File: py/subway.py
import pygame
from pygame.math import Vector2
from agent import Agent
from wall import Wall
from physics import update_agent
import utils
import random

CONFIG = {
    "COUNT": 20,
    "RADIUS": 3,
    "MAXSPEED": 15,
    "MAXFORCE": 50,
    "HORIZON": 7.5,
    "K": 5,
    "AVOID": 15,
    "SIDESTEP": 15,
}

SCREEN_WIDTH = 600
SCREEN_HEIGHT = 600
TIMESTEP = 0.05

pygame.init()
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
clock = pygame.time.Clock()

agents = []
walls = []
positions = {}
frame = 0
assigned = []
animation_done = False

gaps = [-33, -11, 11, 33]

def nearest_door(z):
    return min(gaps, key=lambda g: abs(z - g))

def to_pygame(pos3):
    x, z = pos3[0], pos3[1]
    return Vector2((x + 100) * 3, (z + 100) * 3)

wall_data = [
    (100, False, Vector2(50, 0)),
    (15, True, Vector2(42.5, -50)),
    (15, True, Vector2(42.5, 50)),
    (12, False, Vector2(35, -44)),
    (12, False, Vector2(35, -22)),
    (12, False, Vector2(35, 0)),
    (12, False, Vector2(35, 22)),
    (12, False, Vector2(35, 44)),
]

wall_targets = []
for width, vertical, pos in wall_data:
    start_pos = to_pygame((pos.x, pos.y + 200))
    walls.append(Wall(width * 3, vertical, start_pos))
    wall_targets.append(to_pygame((pos.x, pos.y)))

agents_targets = []

for i in range(CONFIG["COUNT"]):
    pos = utils.get_position(40, 47.5, -39, 39)
    start_pos = to_pygame((pos[0], pos[1] + 200))
    agent = Agent(i, start_pos, CONFIG["RADIUS"], random.uniform(5, CONFIG["MAXSPEED"]),
                  CONFIG["MAXFORCE"], CONFIG["HORIZON"], CONFIG["K"], CONFIG["AVOID"], CONFIG["SIDESTEP"])
    agent.group = 1
    agent.target = start_pos.copy()
    agents.append(agent)
    agents_targets.append(to_pygame(pos))
    assigned.append(False)

for i in range(CONFIG["COUNT"]):
    pos = utils.get_position(0, 23, -39, 39)
    agent = Agent(i + CONFIG["COUNT"], to_pygame(pos), CONFIG["RADIUS"], random.uniform(5, CONFIG["MAXSPEED"]),
                  CONFIG["MAXFORCE"], CONFIG["HORIZON"], CONFIG["K"], CONFIG["AVOID"], CONFIG["SIDESTEP"])
    agent.group = 2
    agent.target = agent.position.copy()
    agents.append(agent)
    assigned.append(False)

running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    if not animation_done:
        done = True
        for i, wall in enumerate(walls):
            wall.float_pos = wall.float_pos.lerp(wall_targets[i], 0.1)
            wall.update_center()
            if (wall.float_pos - wall_targets[i]).length() > 1:
                done = False

        for i, agent in enumerate(agents[:CONFIG["COUNT"]]):
            agent.position = agent.position.lerp(agents_targets[i], 0.1)
            if (agent.position - agents_targets[i]).length() > 1:
                done = False
            agent.target = agents_targets[i].copy()

        if done:
            animation_done = True
            assigned = [False] * len(agents)
    else:
        for agent in agents:
            if not assigned[agent.id]:
                if agent.group == 1:
                    door_y = nearest_door((agent.position.y / 3) - 100)
                    agent.target = to_pygame((25, door_y))
                else:
                    door_y = nearest_door((agent.position.y / 3) - 100)
                    agent.target = to_pygame((42, door_y))
                    agent.phase = "door"
                assigned[agent.id] = True
            else:
                if agent.group == 1:
                    if (agent.target - agent.position).length() < 5:
                        agent.target.x -= 50
                else:
                    if agent.phase == "door" and (agent.target - agent.position).length() < 5:
                        agent.target = to_pygame(utils.get_position(40, 47.5, -39, 39))
                        agent.phase = "inside"

        for agent in agents:
            update_agent(agent, agents, TIMESTEP)

        for agent in agents:
            for wall in walls:
                wall.collision_resolve(agent)

    frame += 1
    positions[frame] = {agent.id: (agent.position.x, agent.position.y) for agent in agents}

    screen.fill((30, 30, 30))

    for wall in walls:
        pygame.draw.rect(screen, (200, 200, 200), wall.rect)

    for agent in agents:
        color = (0, 255, 0) if agent.group == 1 else (255, 0, 0)
        pygame.draw.circle(screen, color, (int(agent.position.x), int(agent.position.y)), int(agent.radius))

    pygame.display.flip()
    clock.tick(60)

pygame.quit()